### PR TITLE
feat(kuma-dp): Kuma DP + Envoy version in Dataplane Insights (#1112)

### PR DESCRIPTION
**Pull request by jewertow**
Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

### Summary

This PR resolves issue #1112.

### Full changelog

* Implement reading envoy version on Kuma DP startup.
* Send Kuma DP and Envoy version in bootstrap request.
* Generate bootstrap with KumaDP and Envoy version.
* Store version of Envoy and Kuma sent in `DiscoveryRequest.Node.Metadata`

### CODE REVIEW

File: app/kuma-dp/pkg/dataplane/envoy/envoy.go [hunk 1]
```diff
@@ -96,7 +103,11 @@ func lookupEnvoyPath(configuredPath string) (string, error) {
 
 func (e *Envoy) Start(stop <-chan struct{}) error {
 	runLog.Info("generating bootstrap configuration")
-	bootstrapConfig, err := e.opts.Generator(e.opts.Config.ControlPlane.URL, e.opts.Config, e.opts.Dataplane)
+	envoyVersion, err := e.version()
+	if err != nil {
+		return errors.Errorf("Failed to get Envoy version. %v", err)
+	}
+	bootstrapConfig, err := e.opts.Generator(e.opts.Config.ControlPlane.URL, e.opts.Config, e.opts.Dataplane, *envoyVersion)
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/pkg/dataplane/envoy/envoy.go:110]:
```
	envoyVersion, err := e.version()
	if err != nil {
		return errors.Wrap(err, "failed to get Envoy version)
	}
	runLog.Info("fetched Envoy version", "version", envoyVersion)
	runLog.Info("generating bootstrap configuration")
	bootstrapConfig, err := e.opts.Generator(e.opts.Config.ControlPlane.URL, e.opts.Config, e.opts.Dataplane, *envoyVersion)

```
1) Slightly different order of logs
2) I think it's useful to add a log of the Envoy version
3) Errors in Go are lowercase
4) Use `errors.Wrap` when you are adding context to errors

File: app/kuma-dp/pkg/dataplane/envoy/envoy.go [hunk 2]
```diff
@@ -156,3 +167,27 @@ func (e *Envoy) Start(stop <-chan struct{}) error {
...
+	if err != nil {
+		return nil, err
+	}
+	arg := "--version"
+	command := exec.Command(resolvedPath, arg)
+	runLog.Info("getting version of Envoy")
+	output, err := command.Output()
+	if err != nil {
+		runLog.Error(err, fmt.Sprintf("the envoy excutable was found at %s but an error occurred when executing it with arg %s", resolvedPath, arg))
+		return nil, err
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/pkg/dataplane/envoy/envoy.go:183]: In Go, you either log or return an error. The only exception for our codebase is `run.go` of binaries. Here - wrap an error and return it

File: app/kuma-dp/pkg/dataplane/envoy/envoy.go [hunk 3]
```diff
@@ -156,3 +167,27 @@ func (e *Envoy) Start(stop <-chan struct{}) error {
...
+
+func (e *Envoy) version() (*EnvoyVersion, error) {
+	binaryPathConfig := e.opts.Config.DataplaneRuntime.BinaryPath
+	resolvedPath, err := lookupEnvoyPath(binaryPathConfig)
+	if err != nil {
+		return nil, err
+	}
+	arg := "--version"
+	command := exec.Command(resolvedPath, arg)
+	runLog.Info("getting version of Envoy")
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/pkg/dataplane/envoy/envoy.go:179]: I think this log is unnecessary. If everything works we will see `fetched envoy version` if it fails we will see `failed to get Envoy version`

File: app/kuma-dp/pkg/dataplane/envoy/remote_bootstrap_test.go [hunk 4]
```diff
@@ -74,15 +87,26 @@ var _ = Describe("Remote Bootstrap", func() {
...
-					expectedBootstrapRequest: `
-                    {
-                      "mesh": "demo",
-                      "name": "sample",
-                      "adminPort": 4321,
-                      "dataplaneTokenPath": "/tmp/token",
-                      "dataplaneResource": "{\"type\":\"Dataplane\",\"mesh\":\"demo\",\"name\":\"sample\",\"creationTime\":\"0001-01-01T00:00:00Z\",\"modificationTime\":\"0001-01-01T00:00:00Z\"}"
-                    }
-`,
+					expectedBootstrapRequest: flatJson(`
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/pkg/dataplane/envoy/remote_bootstrap_test.go:90]: Why do we need `flatJson`? I was pretty sure `MatchJSON` handle new lines etc.

**Review by jakubdyszkiewicz — COMMENTED** [review]
Fantastic work! 👏 

I have a couple of nits and questions but overall really nice.

The next step is now to use this info and put it into DataplaneInsights

**Comment by jakubdyszkiewicz** [comment]
run `make check` to cleanup the code.

Don't worry about failing E2E or minikube tests for now. It's on our side

File: app/kuma-dp/pkg/dataplane/envoy/remote_bootstrap_test.go (hunk 4 above)
Comment (jewertow) [reply app/kuma-dp/pkg/dataplane/envoy/remote_bootstrap_test.go]: You're right. It is redundant and I removed this function.

File: app/kuma-dp/pkg/dataplane/envoy/envoy.go (hunk 1 above)
Comment (jewertow) [reply app/kuma-dp/pkg/dataplane/envoy/envoy.go:112]: Thank you. I fixed it.

File: app/kuma-dp/pkg/dataplane/envoy/envoy.go (hunk 3 above)
Comment (jewertow) [reply app/kuma-dp/pkg/dataplane/envoy/envoy.go]: Ok, I removed this log.

File: app/kuma-dp/pkg/dataplane/envoy/envoy.go (hunk 2 above)
Comment (jewertow) [reply app/kuma-dp/pkg/dataplane/envoy/envoy.go]: Thank you, I fixed it.

**Comment by jewertow** [comment]
Thank you very much for the code review. I fixed imports and the step `make check` passes again.
I will continue work on this PR this week.

**Comment by jewertow** [comment]
I need your help @jakubdyszkiewicz.
I have read the [documentation](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/discovery.proto#envoy-api-msg-discoveryrequest) of `DiscoveryRequest` and I am not sure which property of `DiscoveryRequest` is a proper place for version.

**Comment by jakubdyszkiewicz** [comment]
Hey @jewertow let's do `node.metadata`. I looked into Envoy code and don't really see any references for `user_agent_build_version` in the code so I don't see the benefit of fighting with semver parsing etc.

File: pkg/xds/server/dataplane_status_tracker.go [hunk 5]
```diff
@@ -120,6 +121,9 @@ func (c *dataplaneStatusTracker) OnStreamRequest(streamID int64, req *envoy.Disc
 	if state.dataplaneId == (core_model.ResourceKey{}) {
 		if id, err := core_xds.ParseProxyId(req.Node); err == nil {
 			state.dataplaneId = core_model.ResourceKey{Mesh: id.Mesh, Name: id.Name}
+			if err := util_proto.ToTyped(req.Node.Metadata, state.subscription.Version); err != nil {
+				xdsServerLog.Error(err, "failed to parse Node.Metadata out of DiscoveryRequest", "streamid", streamID, "req", req)
+			}
```
Comment (jakubdyszkiewicz) [inline pkg/xds/server/dataplane_status_tracker.go:126]: what will happen if Envoy won't send the version in the metadata? Will this throw an error or will this just go through? It should be the second option.

also

```
failed to parse Node.Metadata out of DiscoveryRequest
```
->
```
xdsServerLog.Error(err, "failed to extract version out of the Envoy metadata", "streamid", streamID, "metadata", req.Node.Metadata)
```

File: pkg/xds/server/dataplane_status_tracker_test.go [hunk 6]
```diff
@@ -183,7 +206,8 @@ var _ = Describe("DataplaneStatusTracker", func() {
 			// when
 			discoveryRequest := &envoy.DiscoveryRequest{
 				Node: &envoy_core.Node{
-					Id: "default.example-001",
+					Id:       "default.example-001",
+					Metadata: version,
```
Comment (jakubdyszkiewicz) [inline pkg/xds/server/dataplane_status_tracker_test.go:210]: Could you set this a little bit more manually?
```
			Metadata: &pstruct.Struct{
				Fields: map[string]*pstruct.Value{
					"dataplaneTokenPath": &pstruct.Value{
						Kind: &pstruct.Value_StringValue{
							StringValue: "/tmp/token",
						},
					},
				},
			},
```
reference pkg/xds/server/dataplane_metadata_tracker_test.go:19

To be honest I do not fully trust that we have a "version" branch with such serializing

File: pkg/xds/server/dataplane_status_tracker.go (hunk 5 above)
Comment (jewertow) [reply pkg/xds/server/dataplane_status_tracker.go:128]: I tested the following setup:
1. kuma-cp (this branch)
2. kuma-dp (this branch) - sends version
3. kuma-dp (master) - doesn't send version
and there was no failure. Additionally I added unit tests that parse nil and random string and these cases also don't throw error. To be honest I don't know how can I simulate a scenario when it fails... Do you have any idea?

File: pkg/xds/server/dataplane_status_tracker_test.go (hunk 6 above)
Comment (jewertow) [reply pkg/xds/server/dataplane_status_tracker_test.go]: You're right. It didn't work. Thank you. I fixed it and added more test cases.

**Comment by jewertow** [comment]
@jakubdyszkiewicz how command `kumactl inspect dataplanes` should work? 
Should it always return 6 additional columns? Maybe it would be better to add `-version/--version` flag to this command and return columns with versions when the flag is passed?

File: pkg/xds/server/dataplane_status_tracker.go (hunk 5 above)
Comment (jakubdyszkiewicz) [reply pkg/xds/server/dataplane_status_tracker.go:128]: nil / not nil scenario should be fine. Thanks for the extra test!

File: app/kuma-dp/pkg/dataplane/envoy/envoy.go [hunk 7]
```diff
@@ -96,7 +103,13 @@ func lookupEnvoyPath(configuredPath string) (string, error) {
 
 func (e *Envoy) Start(stop <-chan struct{}) error {
 	runLog.Info("generating bootstrap configuration")
-	bootstrapConfig, err := e.opts.Generator(e.opts.Config.ControlPlane.URL, e.opts.Config, e.opts.Dataplane)
+	envoyVersion, err := e.version()
+	if err != nil {
+		return errors.Wrap(err, "failed to get Envoy version")
+	}
+	runLog.Info("fetched Envoy version", "version", envoyVersion)
+	runLog.Info("generating bootstrap configuration")
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/pkg/dataplane/envoy/envoy.go:111]: you forgot to remove the same log in 105

File: api/mesh/v1alpha1/dataplane_insight.proto [hunk 8]
```diff
@@ -93,3 +96,20 @@ message DiscoveryServiceStats {
   // Number of xDS responses NACKed by the Dataplane.
   uint64 responses_rejected = 3;
 }
+
+message Version {
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/dataplane_insight.proto:100]: please comment the fields to follow the rest of the .proto convention

**Comment by jakubdyszkiewicz** [comment]
@jewertow you brought a good question about CLI. Let's do this a separate PR and let me create an issue to discuss.

**Comment by jakubdyszkiewicz** [comment]
please merge origin/master. CI should be green then

File: app/kuma-dp/pkg/dataplane/envoy/envoy.go (hunk 7 above)
Comment (jewertow) [reply app/kuma-dp/pkg/dataplane/envoy/envoy.go:110]: Ok, I removed this redundant log.

File: api/mesh/v1alpha1/dataplane_insight.proto (hunk 8 above)
Comment (jewertow) [reply api/mesh/v1alpha1/dataplane_insight.proto:101]: Ok, I added comments.

**Comment by nickolaev** [comment]
@jewertow would you mind merging `master` please? We will consider including this one after we get 1.0.2 released today.

Also, please do to put WIP in the name of the PR, being a `draft` is already good enough.

**Comment by jewertow** [comment]
@nickolaev I merged master into this branch.

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**